omegaconf/basecontainer.py

```diff
@@ -769,3 +778,39 @@ def _update_types(node: Node, ref_type: type, object_type: Optional[type]) -> No
         if new_ref_type is not Any:
             node._metadata.ref_type = new_ref_type
             node._metadata.optional = new_is_optional
+
+
+def _instantiate_structured_config_impl(
+    conf: "DictConfig", instance_data: Dict[str, Any]
+) -> Any:
+    """Instantiate an instance of `conf._metadata.object_type`, populated by `instance_data`."""
+    from ._utils import get_structured_config_field_names
+
+    object_type = conf._metadata.object_type
+    object_type_field_names = set(get_structured_config_field_names(object_type))
+
+    field_items: Dict[str, Any] = {}
+    nonfield_items: Dict[str, Any] = {}
+    for k, v in instance_data.items():
+        if _is_missing_literal(v):
+            conf._format_and_raise(
+                key=k,
+                value=None,
+                cause=MissingMandatoryValue(
+                    "Structured config of type `$OBJECT_TYPE` has missing mandatory value: $KEY"
+                ),
+            )
+        if k in object_type_field_names:
+            field_items[k] = v
+        else:
+            nonfield_items[k] = v
+
+    result = object_type(**field_items)
+    if not issubclass(object_type, dict):
+        # normal structured config
+        for k, v in nonfield_items.items():
+            setattr(result, k, v)
+    else:
+        # Extending dict as a subclass
+        result.update(nonfield_items)
```

Comment:
You can also force fields into non-dict subclasses with open_dict():

```python
with open_dict(cfg.user):
  cfg.user.root = True
```

dataclasses and attr classes supports adding additional fields. Maybe we should just allow additional fields in the same way?

```python
In [1]: from dataclasses import dataclass

In [2]: @dataclass
   ...: class User:
   ...:     name: str
   ...:

In [3]: user = User(name="joe")

In [4]: user.age = 10

In [5]: user
Out[5]: User(name='joe')

In [6]: user.age
Out[6]: 10

In [7]: import attr

In [8]: @attr.s(auto_attribs=True)
   ...: class User2:
   ...:     name: str
   ...:

In [9]: user2 = User2(name="Bob")

In [10]: user2.age = 11

In [11]: user2
Out[11]: User2(name='Bob')

In [12]: user2.age
Out[12]: 11
```

Reply:
This point is waiting for your input here ;) https://github.com/omry/omegaconf/pull/502#discussion_r594276220

Reply:
As of commit https://github.com/omry/omegaconf/pull/502/commits/3fef7f0563bb1b304bf625e0a54a5dd42b3f9cfd, dict subclass items become instance attributes.